src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -2596,5 +2596,197 @@ public void M(string s)
+                // (25,18): error CS0029: Cannot implicitly convert type 'E' to 'D'
+                //             0 => new E(), // 3.1
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new E()").WithArguments("E", "D").WithLocation(25, 18),
+                // (26,18): error CS0029: Cannot implicitly convert type 'F' to 'D'
+                //             1 => new F(), // 3.2
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new F()").WithArguments("F", "D").WithLocation(26, 18),
+                // (63,18): error CS0029: Cannot implicitly convert type 'E' to 'D'
+                //             0 => new E(), // 9.1
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new E()").WithArguments("E", "D").WithLocation(63, 18),
+                // (64,18): error CS0029: Cannot implicitly convert type 'F' to 'D'
+                //             1 => new F(), // 9.2
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new F()").WithArguments("F", "D").WithLocation(64, 18)
+                );
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var switches = tree.GetRoot().DescendantNodes().OfType<SwitchExpressionSyntax>().ToArray();
+            for (int i = 0; i < switches.Length; i++)
+            {
+                var @switch = switches[i];
+                var typeInfo = model.GetTypeInfo(@switch);
+                var conversion = model.GetConversion(@switch);
+                switch (i)
+                {
+                    case 0:
+                    case 7:
+                        Assert.Null(typeInfo.Type);
+                        Assert.Equal("C", typeInfo.ConvertedType.ToTestDisplayString());
+                        Assert.Equal(ConversionKind.SwitchExpression, conversion.Kind);
+                        break;
+                    case 1:
+                        Assert.Null(typeInfo.Type);
+                        Assert.True(typeInfo.ConvertedType.IsErrorType());
+                        Assert.Equal(ConversionKind.Identity, conversion.Kind);
+                        break;
+                    case 2:
+                    case 8:
+                        Assert.Null(typeInfo.Type);
+                        Assert.Equal("D", typeInfo.ConvertedType.ToTestDisplayString());
+                        Assert.Equal(ConversionKind.SwitchExpression, conversion.Kind);
+                        break;
+                    case 3:
+                        Assert.True(typeInfo.Type.IsErrorType());
+                        Assert.Equal("D", typeInfo.ConvertedType.ToTestDisplayString());
+                        Assert.Equal(ConversionKind.NoConversion, conversion.Kind);
+                        break;
+                    case 9:
+                        Assert.Null(typeInfo.Type);
+                        Assert.Equal("?", typeInfo.ConvertedType.ToTestDisplayString());
```

Comment:
>Assert.Equal("?", typeInfo.ConvertedType.ToTestDisplayString()); [](start = 24, length = 64)

This looks unexpected for the explicit cast case.

Reply:
This is an error case; the switch expression has no natural type, and there is no switch expression conversion applicable.

---
In reply to: [440449065](https://github.com/dotnet/roslyn/pull/45069#discussion_r440449065) [](ancestors = 440449065)